calour/study.py

```diff
@@ -0,0 +1,138 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--,  Calour development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from logging import getLogger
+from os.path import join
+
+import pandas as pd
+import numpy as np
+import scipy
+import biom
+
+
+logger = getLogger(__name__)
+
+
+class Study:
+    '''This class contains the data for a study or a meta study.
+
+    The data set includes a data table (otu table, gene table,
+    metabolomic table, or all those tables combined), a sample
+    metadata table, and a feature metadata.
+
+    Parameters
+    ----------
+    data : ``numpy.array`` or ``scipy.sparse``
+
+    sample_metadata : ``pandas.DataFrame``
+
+    feature_metadata : ``pandas.DataFrame``
+    '''
+    def __init__(self, data, sample_metadata, feature_metadata=None,
+                 description='', sparse=False):
+        self.data = data
+        self.sample_metadata = sample_metadata
+        self.feature_metadata = feature_metadata
+        self.description = description
+
+    @staticmethod
+    def _read_biom(fp, transpose=True, sparse=False):
+        '''Read in a biom table file.
+
+        Parameters
+        ----------
+        fp : str
+            file path to the biom table
+        transpose : bool
+            Transpose the table or not. The OTU table has samples in
+            column while sklearn and other packages require samples in
+            row. So you should transpose the data table.
+        '''
+        table = biom.load_table(fp)
+        sid = table.ids(axis='sample')
+        oid = table.ids(axis='observation')
+        if sparse:
+            data = scipy.sparse.csr_matrix(table.matrix_data)
+        else:
+            data = table.matrix_data.todense().A
```

Comment:
regarding avoiding sparse, some methods are faster on dense arrays, so sometimes useful to store as dense to begin with.
but agree can change default to sparse